### PR TITLE
Set the last three digit of the AMP version to be the number of cherry-picks in the branch

### LINKED ISSUE #27793: I2I: Maintaining AMP version numbers for cherry-picked releases
# Background

## Version numbers

The last significant change to how we determine a release's version number was #16631. In that change we switched from UNIX timestamp (13 digits) to a date-time based solution (format YYmmDDHHMMSS) (12 digits) - to ensure the version numbers are monotonically increasing we added an arbitrary digit X at the end of the version (YYmmDDHHMMSSX).

Previously this digit was used to denote whether the version was built from a "clean workspace" (0 - no uncommitted changes; 1 - there are uncommitted changes) but this proved to not be useful, so we changed the version number to just default X to 0, and have it arbitrarily changeable by passing a digit to `gulp dist --custom_version_mark` (now `--custom_version` which sets the entire version number). This feature was used exactly once.

## Cherry picks

When a breaking issue (one that meets the [cherry pick](https://github.com/ampproject/amphtml/blob/master/contributing/contributing-code.md#Cherry-picks) criteria) is discovered on Stable or other channel with active traffic to them we rollback to a previously known good version, create and merge a fix to master, then cherry pick the fix onto the broken version, and deploy it. (on rare occasions where the fix is already available, we sometimes skip the rollback and deploy directly on top of the breaking release. This is done at the discretion of the TSC)

The version number resulting in this cherry pick uses the time when the PR that fixes the issue is _cherry picked_ onto the existing release. While in line with the method chosen in #16631, this creates difficulties in tracing back which versions are cherry picked versions and on top of which release they were created.

For example, the following is a graph of some recent releases:

graph ID #b45861bf96eabd9eebfc4201a7fdb195

###### Releases marked in blue were built from commits that are on the `master` branch; commits in black were created with cherry picked commits. _NOTE: this graph was modified slightly from the reality! This is due to a combination of changes in the release process **and** human error during the release, that would have made this solution seem much more complicated than it would be in reality._

While investigating a regression that was discovered on 2004142326360 (the release marked as `Stable (rolled back)`, and was not present in 2004030010070 (marked as `Stable` - the Stable version previous to the one that was rolled back) it was difficult for AMP developers to perform a bisection of the releases and find the root cause. This convenient graph was not initially available and had to be generated manually. (for those curious about the issue that caused all this trouble, the rabbit-hole starts here: #27790)

Unlike releases built from commits that are directly on `master`, that correspond to a specific time and date in the revision history of this project, the cherry picked releases do not help trace back where their code lives or what versions came prior. They only indicate when the cherry-pick was performed; this does not expose relevant information.

## Proposal

The proposal here is to use the trailing digit as an indicator of the number of cherry picks on top of an existing release. e.g., the first cherry pick on top of `2003241730220` will be called `2003241730221`. A second cherry pick will be called `2003241730222`, and so on.

If there are more than 9 cherry-picks, a new release will have to be performed. Since we switched to the current versioning schema, there have been 6 releases (on 3 release branches) that required more than 9 cherry-picks, and only 1 such in the past year.

Using this new versioning schema, the above graph would have looked like this:

graph ID #38cece0dcfc7e786b6d426ee2390a255

/cc @ampproject/wg-approvers @ampproject/wg-infra

### PR DESCRIPTION
**Pull request by danielrozenberg**
Implements and closes #27793

### CODE REVIEW

File: build-system/compile/internal-version.js [hunk 1]
```diff
@@ -26,40 +26,75 @@ const argv = minimist(process.argv.slice(2), {
...
+  // - Find the commit (<C>) before the last cherry-picked commit from the
+  //   `master` branch (if the current branch is `master`, or otherwise in
+  //   `master`'s commit history, then the current commit is <C>).
+  // - Find the commit time of <C> (<C>.time). Note that commit time might be
+  //   different from author time! e.g., commit time might be the time that a PR
+  //   was merged into `master`, or a commit was cherry-picked onto the brabnch;
+  //   author time is when the original author of the commit ran the
+  //   "git commit" command.
+  // - The version number is <C>.time.format("YYmmDDHHMMSS", "UTC") + <X> (the
+  //   pseudo-code assumes standard `.strftime` formatting).
```
Comment (rcebulko) [inline build-system/compile/internal-version.js:50]: Would it be easier/clearer to speak relative to the RTV of the main release that's being cherry-picked, rather than the commit time?

File: build-system/compile/internal-version.js [hunk 2]
```diff
@@ -26,40 +26,75 @@ const argv = minimist(process.argv.slice(2), {
...
+  //      last of which was commited on November 10, 2021 5:01:12 PM PST,
+  //    - followed by four commits that were cherry-picked from `master`,
+  //    then its version number would be `202111110101124`.
+  //    - The latest four commits are cherry-picks from `master`, and the one
+  //      before them is not, so our last digit is set to 4.
+  //    - PST is UTC-8, so the hour value changes from PST's 17 to UTC's 1, and
+  //      one day is added.
+  let numberOfCherryPicks = 0;
+  const commitCherriesInfo = gitCherryMaster().reverse();
+  for (const commitCherryInfo of commitCherriesInfo) {
```
Comment (rcebulko) [inline build-system/compile/internal-version.js:81]: Nit: destructure
```suggestion
  for (const {isCherryPick} of commitCherriesInfo) {
```

File: build-system/compile/internal-version.js [hunk 3]
```diff
@@ -26,40 +26,75 @@ const argv = minimist(process.argv.slice(2), {
...
+  const lastCommitFormattedTime = gitCommitFormattedTime(
+    `HEAD~${numberOfCherryPicks}`
+  );
+
+  if (argv.strict_build && numberOfCherryPicks > 9) {
+    throw new Error(
+      `This branch has ${numberOfCherryPicks} cherry-picks. --strict_build caps the number of cherry-picks at 10. To build this branch use --version_override.`
+    );
   }
+  const lastDigit = Math.min(numberOfCherryPicks, 9);
```
Comment (rcebulko) [inline build-system/compile/internal-version.js:96]: Nit: Seems unlikely to change until the next time we change versioning somehow, but I feel like `9` should be a constant `MAX_CHERRY_PICKS` or something

File: build-system/compile/internal-version.js [hunk 4]
```diff
@@ -26,40 +26,75 @@ const argv = minimist(process.argv.slice(2), {
...
+  //      one day is added.
+  let numberOfCherryPicks = 0;
+  const commitCherriesInfo = gitCherryMaster().reverse();
+  for (const commitCherryInfo of commitCherriesInfo) {
+    if (!commitCherryInfo.isCherryPick) {
+      break;
     }
-    return `${lastCommitFormattedTime}${lastDigit}`;
+    numberOfCherryPicks++;
+  }
```
Comment (rcebulko) [inline build-system/compile/internal-version.js:86]: Is it important that the above loop short-circuit? It should be impossible for there to be a non-cherry-pick between two cherry-pick commits, correct? If that's the case, lines 79-86 could become

```javascript
const numCherryPicks = gitCherryMaster().filter(({isCherryPick}) => isCherryPick).length;
```

File: build-system/compile/internal-version.js (hunk 3 above)
Comment (rcebulko) [reply build-system/compile/internal-version.js:96]: Also, why do we want to allow >9 for non-strict builds?

File: build-system/compile/internal-version.js (hunk 1 above)
Comment (danielrozenberg) [reply build-system/compile/internal-version.js:50]: This is what this PR does :D
RTV is made our of the commit time (see #16631)

File: build-system/compile/internal-version.js (hunk 3 above)
Comment (danielrozenberg) [reply build-system/compile/internal-version.js:96]: 9 being the highest digit in our counting system, I feel like this is one of those exceptions to making it a const. Don't feel super strongly about this though.

The idea I had is that strict builds are what we create for releases - developers might have cherry-picks, regular commits, more cherry picks, etc - not failing their local builds seems like a good choice. If it all ends up being a PR it'll be squashed to a single commit at the end anyways and wouldn't interfere with the release process

File: build-system/compile/internal-version.js (hunk 4 above)
Comment (danielrozenberg) [reply build-system/compile/internal-version.js:86]: Yes, because of the local workspace for developers scenario

File: build-system/compile/internal-version.js (hunk 2 above)
Comment (danielrozenberg) [reply build-system/compile/internal-version.js]: Neat! Pushed it myself to avoid the annoying CLA bot issue

File: build-system/common/git.js [hunk 5]
```diff
@@ -169,12 +169,30 @@ function gitCommitterEmail() {
...
+ * Returns list of commit SHAs and their cherry-pick status from master.
+ *
+ * @return {!Array<{sha: string, isCherryPick: boolean}>}
+ */
+function gitCherryMaster() {
+  return getStdout('git cherry master')
+    .trim()
+    .split('\n')
+    .map((line) => ({
+      isCherryPick: line.substring(0, 2) == '- ',
```
Comment (rsimha) [inline build-system/common/git.js:181]: This can be:
```js
isCherryPick: line.startsWith('- '),
```

File: build-system/common/git.js [hunk 6]
```diff
@@ -169,12 +169,30 @@ function gitCommitterEmail() {
 }
 
 /**
- * Returns the timestamp of the latest commit on the local branch.
+ * Returns list of commit SHAs and their cherry-pick status from master.
```
Comment (rsimha) [inline build-system/common/git.js:172]: Expand this JSDoc description to briefly mention that `- <sha>` indicates a cherry-picked commit while `+ <sha>` indicates a new commit on the branch.

File: build-system/common/git.js [hunk 7]
```diff
@@ -169,12 +169,30 @@ function gitCommitterEmail() {
...
+
+/**
+ * Returns the timestamp of a commit on the local branch.
+ *
+ * @param {string} ref a Git reference (commit SHA, branch name, etc.) for the
+ *   commit to get the time of. Leave out or set to empty string for HEAD.
  * @return {number}
  */
-function gitCommitFormattedTime() {
+function gitCommitFormattedTime(ref = '') {
```
Comment (rsimha) [inline build-system/common/git.js:193]: Would it be more explicit to set the default argument to `ref = 'HEAD'`?

File: build-system/compile/internal-version.js [hunk 8]
```diff
@@ -26,40 +26,75 @@ const argv = minimist(process.argv.slice(2), {
...
-      lastDigit = parseInt(argv.custom_version_mark, 10);
-      if (isNaN(lastDigit) || lastDigit < 0 || lastDigit > 9) {
-        throw new Error(
-          `--custom_version_mark is set to ${argv.custom_version_mark}, ` +
-            'expected value between 0 and 9!'
-        );
-      }
+  }
+  // Version numbers are determined using the following algorithm:
+  // - Count the number (<X>) of cherry-picked releases on this branch that came
```
Comment (rsimha) [inline build-system/compile/internal-version.js:35]: nit for accuracy: "commits", not "releases", right?

> Count the number (\<X\>) of cherry-picked _**commits**_ on this branch that...

File: build-system/compile/internal-version.js [hunk 9]
```diff
@@ -26,40 +26,75 @@ const argv = minimist(process.argv.slice(2), {
...
+      break;
     }
-    return `${lastCommitFormattedTime}${lastDigit}`;
+    numberOfCherryPicks++;
+  }
+  const lastCommitFormattedTime = gitCommitFormattedTime(
+    `HEAD~${numberOfCherryPicks}`
+  );
+
+  if (argv.strict_build && numberOfCherryPicks > 9) {
```
Comment (rsimha) [inline build-system/compile/internal-version.js:91]: usability nit: `--strict_build` didn't convey what this meant until I read all the comments. How about `--verify_cherry_picks`? (Also, see my proposal below to avoid adding this flag altogether.)

File: build-system/compile/internal-version.js (hunk 3 above)
Comment (rsimha) [reply build-system/compile/internal-version.js:96]: > I feel like 9 should be a constant `MAX_CHERRY_PICKS` or something

I agree that `9` here and on line 91 should be hoisted to a const at the top of this file.

> The idea I had is that strict builds are what we create for releases

Reopening this thread to further play out this scenario. I do see the rationale behind adding a flag to cap the cherry-pick count at 9. However, I don't see a graceful way to recover from the case where there are more than 9 cherry-picked commits in a given release branch (however unlikely it is in practice), and the strict cap is enforced.

- If we do enforce the cap of 9 CPs, we can't do any more releases from that branch and will have to cut a new one (correct?)
- If we don't enforce the cap, and there are two releases branches, one with 9 CPs and another with 10, they'll have the same version, which is also undesirable

I don't know a great solution to this with a single digit for CP status. In addition, I'm not super enthusiastic about adding yet another flag when `gulp dist` is invoked by the release automation, since maintenance and future flag changes require an elaborate 2 week process to avoid skew between release code and source code.

One alternative solution that comes to mind is to eliminate the two digits for `seconds` from CP version numbers, and use the last three digits to indicate CP status.

Pros as I see them:
- This will work the same for local development and releases
- No need for the additional flag
- CP release branches will be easy to identify and sort. (E.g. For the main branch of `2004161906340`, the first CP branch will be `2004161906000`, the second `2004161906001`, the third `2004161906002`, etc.)

Do you think this is feasible? Any drawbacks I'm not considering?

Comment (danielrozenberg) [reply build-system/compile/internal-version.js:96]: After the discussion on the I2I, went with dropping the seconds

File: build-system/common/git.js (hunk 5 above)
Comment (danielrozenberg) [reply build-system/common/git.js:181]: I went this route because of its relationship to the next line, basically I'm saying "split this at character index 2", I believe this increases the readability in this particular case at not significant additional cost in performance :)

File: build-system/common/git.js (hunk 6 above)
Comment (danielrozenberg) [reply build-system/common/git.js:172]: Done

File: build-system/common/git.js (hunk 7 above)
Comment (danielrozenberg) [reply build-system/common/git.js:193]: Done

File: build-system/compile/internal-version.js (hunk 8 above)
Comment (danielrozenberg) [reply build-system/compile/internal-version.js:35]: Done

File: build-system/compile/internal-version.js (hunk 9 above)
Comment (danielrozenberg) [reply build-system/compile/internal-version.js:91]: Obsolete

File: build-system/compile/internal-version.js [hunk 10]
```diff
@@ -26,40 +26,77 @@ const argv = minimist(process.argv.slice(2), {
...
-    return `${lastCommitFormattedTime}${lastDigit}`;
+    numberOfCherryPicks++;
+  }
+  if (numberOfCherryPicks > 999) {
+    throw new Error(
+      `This branch has ${numberOfCherryPicks} cherry-picks, which is more than 999, the maximum allowed number of cherry-picks!`
+    );
   }
+
+  const lastCommitFormattedTime = gitCommitFormattedTime(
```
Comment (mdmower) [inline build-system/compile/internal-version.js:94]: This is the only place `gitCommitFormattedTime()` is used. Could you revise the date format to omit `%S` and avoid the `.slice(0, -2)`?

Comment (danielrozenberg) [reply build-system/compile/internal-version.js:94]: I thought about it, but I prefer to keep the original in case someone else will use the other function expecting the existing behaviour

File: build-system/compile/internal-version.js [hunk 11]
```diff
@@ -26,40 +26,77 @@ const argv = minimist(process.argv.slice(2), {
...
+  //   from the `master` branch, until reaching `master` or the first commit
+  //   that was added directly on this branch (if the current commit is on
+  //   `master`'s commit history, or only contains new commits that are not
+  //   cherry-picked from `master`, then <X> is 0).
+  // - Find the commit (<C>) before the last cherry-picked commit from the
+  //   `master` branch (if the current branch is `master`, or otherwise in
+  //   `master`'s commit history, then the current commit is <C>).
+  // - Find the commit time of <C> (<C>.time). Note that commit time might be
+  //   different from author time! e.g., commit time might be the time that a PR
+  //   was merged into `master`, or a commit was cherry-picked onto the brabnch;
```
Comment (mdmower) [inline build-system/compile/internal-version.js:45]: Typo: brabnch

File: build-system/compile/internal-version.js [hunk 12]
```diff
@@ -26,40 +26,77 @@ const argv = minimist(process.argv.slice(2), {
...
+  //
+  // 3. For a release built from a local working branch that was split off from
+  //    a `master` commit from November 9, 2021 11:48 PM PST, and then:
+  //    - had one commit that was cherry-picked from `master`,
+  //    - followed by two commits that were created directly on the branch, the
+  //      last of which was commited on November 23, 2021 6:38 PM PST,
+  //    - followed by twelve commits that were cherry-picked from `master`,
+  //    then its version number would be `2111240238014`.
+  //    - The latest twelve commits are cherry-picks from `master`, and the one
+  //      before them is not, so our last three digits are set to 014.
```
Comment (mdmower) [inline build-system/compile/internal-version.js:77]: I'm not following this bullet point. The 12 most recent commits are cherry picks, but how do the two non-cherry-picks plus one earlier cherry-pick end up adding 2 to the original 12? Your introduction makes it sound like this should be `012`.

File: build-system/compile/internal-version.js (hunk 11 above)
Comment (danielrozenberg) [reply build-system/compile/internal-version.js:45]: Done

File: build-system/compile/internal-version.js (hunk 12 above)
Comment (danielrozenberg) [reply build-system/compile/internal-version.js]: Finger slip :D
Fixed, this really should have been 012 as you pointed out

File: build-system/common/git.js [hunk 13]
```diff
@@ -169,12 +169,36 @@ function gitCommitterEmail() {
...
+ * `git cherry <branch>` returns a list of commit SHAs. While the exact
+ * mechanism is too complicated for this comment (run `git help cherry` for a
+ * full explanation), the gist of it is that commits that were cherry-picked
+ * from <branch> are prefixed with '- ', and those that were not are prefixed
+ * with '+ '.
+ *
+ * @return {!Array<{sha: string, isCherryPick: boolean}>}
+ */
+function gitCherryMaster() {
+  return getStdout('git cherry master')
```
Comment (rsimha) [inline build-system/common/git.js:183]: A use case I just thought of: Today, it isn't necessary for developers to keep their local `master` branch in sync with `upstream/master`. For example: 

- A developer has a `developer/amphtml` fork that is a couple of months old
- They create a local branch `foo`
- Sync `foo` to `upstream/master`
- Add a bunch of commits to it
- Meanwhile, the local `master` branch is still two months old

Will this result in the new versioning logic detecting hundreds of "cherry-picked" commits?

Comment (rsimha) [reply build-system/common/git.js:183]: Answering my own question: This isn't a problem because the additional commits between the outdated local `master` and `HEAD` are counted as new commits, not cherry-picked commits. Carry on!

Screen Shot 2020-04-22 at 2 46 10 PM

File: build-system/compile/internal-version.js [hunk 14]
```diff
@@ -26,40 +26,77 @@ const argv = minimist(process.argv.slice(2), {
...
+  //    a `master` commit from November 9, 2021 11:48 PM PST, and then:
+  //    - had one commit that was cherry-picked from `master`,
+  //    - followed by two commits that were created directly on the branch, the
+  //      last of which was commited on November 23, 2021 6:38 PM PST,
+  //    - followed by twelve commits that were cherry-picked from `master`,
+  //    then its version number would be `2111240238012`.
+  //    - The latest twelve commits are cherry-picks from `master`, and the one
+  //      before them is not, so our last three digits are set to 012.
+  //    - PST is UTC-8, so the hour value changes from PST's 18 to UTC's 2, and
+  //      one day is added.
```
Comment (rsimha) [inline build-system/compile/internal-version.js:79]: For better code readability, can we move this to a JSDoc block at the top of this function?

File: build-system/compile/internal-version.js [hunk 15]
```diff
@@ -26,40 +26,77 @@ const argv = minimist(process.argv.slice(2), {
...
+    numberOfCherryPicks++;
+  }
+  if (numberOfCherryPicks > 999) {
+    throw new Error(
+      `This branch has ${numberOfCherryPicks} cherry-picks, which is more than 999, the maximum allowed number of cherry-picks!`
+    );
   }
+
+  const lastCommitFormattedTime = gitCommitFormattedTime(
+    `HEAD~${numberOfCherryPicks}`
```
Comment (rsimha) [inline build-system/compile/internal-version.js:95]: Sanity check: Does this count the number of commits on the CP branch? Or does it count just  the number of cherry-picked commits?

In other words, I think this should count both `+` and `-` prefixed commits in order to get to the branch point, right?

File: build-system/compile/internal-version.js (hunk 14 above)
Comment (rcebulko) [reply build-system/compile/internal-version.js:79]: I would almost support putting this in a markdown doc and linking, this is lengthy :o

File: build-system/compile/internal-version.js (hunk 15 above)
Comment (danielrozenberg) [reply build-system/compile/internal-version.js:95]: I was worried about the case where somebody is restarting an old PR and does a merge commit on their working branch - suddenly there could be thousands of commits in between because of the merge

The method in the git helper module does return all of them, and can be used somewhere else to verify that a release only contains commits directly,or cherry-picked, from `master`

Comment (rsimha) [reply build-system/compile/internal-version.js:95]: My concern is whether we're using the timestamp of the correct commit here. 

Suppose there are 3 cherry-picked commits and 4 new commits in a CP branch. Shouldn't this function use the timestamp of `HEAD~7` and not `HEAD~3`? Or could it be that I'm missing something?

File: build-system/compile/internal-version.js (hunk 14 above)
Comment (danielrozenberg) [reply build-system/compile/internal-version.js]: Done

File: build-system/compile/internal-version.js (hunk 15 above)
Comment (rsimha) [reply build-system/compile/internal-version.js:105]: Just ran some tests on my local branch, and this is indeed the wrong timestamp. (In the example above, it should be `HEAD~7`, not `HEAD~3`.)

File: build-system/common/git.js (hunk 13 above)
Comment (danielrozenberg) [reply build-system/common/git.js:183]: This is exactly why I stop counting once I get to the first not-cherry-pick commit :)

File: build-system/compile/internal-version.js (hunk 15 above)
Comment (danielrozenberg) [reply build-system/compile/internal-version.js:105]: I designed it specifically to use the last "new" commit (aka not a cherry-picked commit) instead of using the branch point.

Imagine a situation where some creates a branch from release `1234567890000` and adds 1 new commit to it. Do we want their version number to *also* be `123456789000`? There are no cherry picks, we want to use the time of the last new commit in the branch for that :)

Comment (danielrozenberg) [reply build-system/compile/internal-version.js:105]: (also this is exactly example 3 in the doc)

File: build-system/compile/internal-version.js (hunk 14 above)
Comment (danielrozenberg) [reply build-system/compile/internal-version.js]: @rcebulko just saw your comment now, after I wrote Done (race conditions yay). I think it's important to detail the algorithm where it's at, especially if it's not obvious from reading the code what it does and how it handles the multiple edge cases.

It also makes sense to document it in some relevant file, in a less detailed fashion. Maybe release-schedule.md? DEVELOPING.md?

Comment (rsimha) [reply build-system/compile/internal-version.js]: JSDoc next to implementation SGTM :)

File: build-system/compile/internal-version.js (hunk 15 above)
Comment (rsimha) [reply build-system/compile/internal-version.js:105]: Thanks for explaining. For some reason, that last point was lost to me. Also, when I tested your branch locally, I had to include your commits, which show up as "new" commits and changed the calculation. 🤦

**Comment by danielrozenberg** [comment]
@mdmower I would love an approval from you before I merge this!

**Comment by mdmower** [comment]
> @mdmower I would love an approval from you before I merge this!

@danielrozenberg I appreciate the consideration! I think the idea is sound, but I have the same concern that @rsimha mentioned [here](https://github.com/ampproject/amphtml/pull/27848#discussion_r413225326). He seems to consider it a non-issue, so I still have to reconcile these results for myself:

```
$ git fetch origin
$ git checkout master
$ git reset --hard 2001232119410 # an old tag that exists on master
$ git checkout 2003262059300 # a recent tag with a cherry pick
$ git cherry master | grep '-'
(no results)
$ git cherry origin/master
- (hash)
```

To me this looks like an out of date local `master` branch would result in an inconsistent version compared to an up-to-date `master`. I might be able to convince myself this isn't a problem this evening (after 5pm PDT), but I need to do paid-job work until then.

**Comment by danielrozenberg** [comment]
No rush!

**Comment by mdmower** [comment]
@danielrozenberg @rsimha I haven't resolved the dilemma I mentioned in my previous comment. Below is a simplified version. The issue I'm seeing is that an out of date local `master` branch could result in a different cherry-pick count _if_ the cherry-picks come from the upstream `master`. The reason for the discrepancy is that this PR references the local `master` branch when evaluating `git cherry master`.

```
$ git fetch origin master:master
$ git checkout master
$ git reset --hard origin/master~1 # local master is 1 commit behind upstream
$ git tag cptest # tag marks where release was cut (1 commit behind upstream)
$ git checkout cptest
$ git cherry-pick origin/master
$ git cherry master
+ 721981a51b52dd83ae8c6e0f2b3af09f290df43d
$ git cherry origin/master
- 721981a51b52dd83ae8c6e0f2b3af09f290df43d
```

If you agree this is a concern, then you could switch to running `git cherry` against the latest remote master:
```
git cherry "$(git config branch.master.remote)/master"
```

**Comment by danielrozenberg** [comment]
This is an amazing deep dive into the different ways one can use Git :)

I think this is mostly a non-issue, as I doubt many people will bother doing a fetch from `origin/master` without fast-forwarding their local `master` branch, _then_ create a tag, and _then_ cherry pick from `origin/master`. Even if they do, in that case their local version number will be the time of their cherry-pick, which isn't a total failure, as it only affects their local workspace.

We enforce (both through automation and through due-process) an invariant that would prevent this from affecting actual releases - we reject releases that would return any `+ ` commits on `git cherry master` (or in a human readable format, the invariant is "reject a release that contains any commits that did not come from `master`)"

I'll merge this PR now so that nightly builds starting next week will use this new format, we can always tweak the finer details later if you still think this is an edge case that should be addressed

**Comment by mdmower** [comment]
> I think this is mostly a non-issue, as I doubt many people will bother doing a fetch from origin/master without fast-forwarding their local master branch, then create a tag, and then cherry pick from origin/master

Just to clarify, I was providing steps for reproducing the issue. They not all necessary in practice. All that's necessary is for someone to run `git fetch origin`. Then, if they check out a recent release with cherry-picks (e.g. `2004240001480`) and build it, it is possible to end up with a different cherry-pick count because their local master was never updated.

**Comment by danielrozenberg** [comment]
I see... Though I don't know why someone would do that :)
In any case, this can only affect your local workspace - if you mess around with git you can always break stuff - we're trying to protect against: release numbers going wrong, and AMP developers not being able to run `gulp build` because they messed around with with - what AMP version number they get on their local machine rarely matters

**Comment by jridgewell** [comment]
I'm with @mdmower here, this can cause all kinds of problems. For instance, I don't frequently update my local `master`. Instead, I do `git checkout -b BRANCH upstream/master`, which creates my new branches directly from whatever `ampproject/amphtml`'s `master` is currently on.

As, is, I will _always_ return the wrong number of cherry-picks.

File: build-system/compile/internal-version.js [hunk 16]
```diff
@@ -16,50 +16,97 @@
...
   }
+  if (numberOfCherryPicks > 999) {
+    throw new Error(
+      `This branch has ${numberOfCherryPicks} cherry-picks, which is more than 999, the maximum allowed number of cherry-picks!`
+    );
+  }
+
+  const lastCommitFormattedTime = gitCommitFormattedTime(
+    `HEAD~${numberOfCherryPicks}`
+  ).slice(0, -2);
```
Comment (jridgewell) [inline build-system/compile/internal-version.js:106]: This seems to be the only caller. Why not just remove the `%S` in the formatted time?

Comment (mdmower) [reply build-system/compile/internal-version.js:106]: https://github.com/ampproject/amphtml/pull/27848#discussion_r412455890

**Comment by danielrozenberg** [comment]
> Per [#27848 (comment)](https://github.com/ampproject/amphtml/pull/27848#issuecomment-619266327), I think this should be reverted.

@jridgewell does it really matter to you what AMP version number shows up when you're doing local development? My main concern with this PR was fixing version numbers for _releases_, while making sure that local development will continue working uninterrupted (i.e., no errors or completely incomprehensible version numbers)

**Comment by mdmower** [comment]
I'm hopeful that #28035 will alleviate concern and this commit can be left in-place (no revert necessary).

**Comment by jridgewell** [comment]
> My main concern with this PR was fixing version numbers for releases, while making sure that local development will continue working uninterrupted

This isn't for local development. Our cherry-pick process boils down to:

```js
git clone git@github.com:ampproject/amphtml.git ampproject/amphtml
cd ampproject/amphtml
git checkout -b amp-release-2020-04-27 ${AMP_VERSION}
git cherry-pick -x ${SHA}
git push --set-upstream origin amp-release-2020-04-27
```

If my local `master` is out of date, then we've got issues when creating the official release.

**Comment by danielrozenberg** [comment]
That's why the first line there is `git clone`, so you'll be working with a fresh copy ;)
@estherkim is working on automating the cherry-picking process directly using GitHub's API, so this would not be a major concern going forward

**Comment by jridgewell** [comment]
Why would I clone the  repo if I already have it? We should add an explicit step to `git checkout -B master upstream/master`, since it's now a requirement.

**Comment by jridgewell** [comment]
I actually don't understand why the cherry-pick logic is looking at `master` at all. Shouldn't we be diffing against the `VERSION` tag that we branched from?

**Comment by estherkim** [comment]
It guarantees that you have the cherry pick commit, which is usually a recently merged PR.

**Comment by jridgewell** [comment]
Can you explain what you mean?

**Comment by estherkim** [comment]
Hmm well I'm not following the "diffing against the `VERSION` tag" part. We create a new branch based off the `VERSION` tag, then apply a commit from `master` on top of that new branch.

**Comment by rsimha** [comment]
@jridgewell sorry to pile on to this discussion, but can you summarize your main objection to this PR? (Is the problem the [originally stated goal](https://github.com/ampproject/amphtml/issues/27793#issue-600638907)? Or is it the `master` [comparison logic](https://github.com/ampproject/amphtml/blob/9ff2ffdd319cccdb21093605d10fa6b7dc02593b/build-system/compile/internal-version.js#L29-L47)?)

> I actually don't understand why the cherry-pick logic is looking at `master` at all. Shouldn't we be diffing against the `VERSION` tag that we branched from?

In order to identify which commits were cherry-picked from `master` to a release branch, we need a view of a) the release branch itself, b) the point at which it branched off from `master`, and c) the up-to-date `master` branch.

- Knowledge of a) and b) is trivial when you are on the release branch.
- Knowledge of c) requires one of these two things:
    1. Update local `master` during the CP process (https://github.com/ampproject/amphtml/pull/27848#issuecomment-620206158)
    2. Compare commits in the CP branch against `upstream/master` instead of local `master` (https://github.com/ampproject/amphtml/pull/27848#issuecomment-620122899) 

See the example below.

image
